src/RA_Defs.cpp

```diff
@@ -9,55 +9,74 @@ GetParseErrorFunc GetJSONParseErrorStr = GetParseError_En;
 
 #endif
 
-static_assert(sizeof(BYTE*) == sizeof(char*), "dangerous cast ahead");
-char* DataStreamAsString(DataStream& stream)
-{
-    return reinterpret_cast<char*>(stream.data());
-}
+namespace ra {
 
-std::string Narrow(const wchar_t* wstr)
+std::string Narrow(const std::wstring& wstr)
 {
-    static std::wstring_convert< std::codecvt_utf8_utf16< wchar_t >, wchar_t > converter;
-    return converter.to_bytes(wstr);
+    return Narrow(wstr.c_str());
 }
 
-std::string Narrow(const std::wstring& wstr)
+std::string Narrow(std::wstring&& wstr) noexcept
 {
-    static std::wstring_convert< std::codecvt_utf8_utf16< wchar_t >, wchar_t > converter;
-    return converter.to_bytes(wstr);
+    auto wwstr{ std::move_if_noexcept(wstr) };
+    return Narrow(wwstr);
 }
 
-std::wstring Widen(const char* str)
+
+std::string Narrow(const wchar_t* wstr)
 {
-    static std::wstring_convert< std::codecvt_utf8_utf16< wchar_t >, wchar_t > converter;
-    return converter.from_bytes(str);
+    auto state{ std::mbstate_t{} };
+    auto len{ 1 + std::wcsrtombs(nullptr, &wstr, 0_z, &state) };
+
+    // doesn't work with unique_ptr, though it works with sprintf and the functions in RAWeb
+    std::string str;
+    str.reserve(len);
+
+    ::WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, wstr, to_signed(len), str.data(),
+        to_signed(str.length()), nullptr, nullptr);
```

Comment:
unit tests caught this one: `str.length()` is 0 at this point. it should be `wcslen(wstr) + 1`

i'm not sure what's going on with the second EARTH GLOBE unit test, `wcsrtombs` is returning -1, I suspect it doesn't support surrogate pairs.

Reply:
what is the rationale behind changing this from `std::wstring_convert`?

Reply:
The JSON responses were getting parsed and I saw the Rich Presence stuff still getting printed correctly. 

Rather than explaining it you can read it.
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0618r0.html

At this point I don't really care too much, it needed to be capacity and not length since the size is 0.

Reply:
Just make sure the unit tests pass. That's why they're there.

Reply:
Thanks for pointing it out. I checked it out using the memory bookmarker for saving files and it was able to write to it in tact.

Reply:
RA_Defs_Tests::TestNarrow is still failing:
```
Message: Assert failed. Expected:<Test> Actual:<TestÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍýýýý>
```
I had to add a +1 to the length to get it to copy the null terminator:
```
    ::WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, wstr, lstrlenW(wstr) + 1, str.data(),
        str.capacity(), nullptr, nullptr);
```
Also changed the `lstrlenW(wstr) * 4` to `str.capacity()` to avoid a second strlen call.